stock.py

```diff
@@ -23,86 +25,78 @@ def __setup__(cls):
-            return
-
-        cost_lines = self._on_change_product_cost_lines()
-        if cost_lines:
-            cost_lines = cost_lines.get('add')
-            LotCostLine = Pool().get('stock.lot.cost_line')
-            lot_cost_lines = LotCostLine.search([
-                    ('lot', '=', self.id),
-                    ('category', '=', cost_lines[0][1]['category']),
-                    ('unit_price', '=', cost_lines[0][1]['unit_price']),
-                    ])
-            if lot_cost_lines:
-                self.cost_lines = lot_cost_lines
-
-    def _on_change_product_cost_lines(self):
+    cost_price = fields.Function(fields.Numeric("Cost Price"),
+        'get_lot_prices')
+    total_cost = fields.Function(fields.Numeric("Total Cost"),
+        'get_lot_prices')
+
+    @classmethod
+    def get_lot_prices(cls, lots, names):
         pool = Pool()
-        ModelData = pool.get('ir.model.data')
-
-        if not self.product:
-            return {}
-
-        category_id = ModelData.get_id('stock_lot_cost',
-            'cost_category_standard_price')
-        return {
-            'add': [(0, {
-                        'category': category_id,
-                        'unit_price': self.product.cost_price,
-                        })],
-            }
+        Move = pool.get('stock.move')
+        Product = pool.get('product.product')
+        Location = pool.get('stock.location')
+        Company = pool.get('company.company')
+
+        res = {}
+        ids = [x.id for x in lots]
+        for name in ['cost_price', 'total_cost']:
+            res[name] = dict.fromkeys(ids)
+
+        warehouse_ids = [location.id for location in Location.search(
+            [('type', '=', 'warehouse')])]
+        product_ids = list(set(lot.product.id for lot in lots))
+        lot_ids = list(set(lot.id for lot in lots))
+
+        moves = Move.search([
+            ('lot', 'in', lot_ids),
+            ('origin', 'not like', 'stock.move,%'),
+            ('from_location.type', '=', 'supplier'),
+            ('to_location.type', '=', 'storage'),
+            ('state', '=', 'done'),
+            ])
+
+        with Transaction().set_context({'stock_date_end': datetime.date.max}):
+            pbl = Product.products_by_location(warehouse_ids,
+                with_childs=True,
+                grouping=('product', 'lot'),
+                grouping_filter=(product_ids, lot_ids))
+
+        lot_moves = {}
+        for move in moves:
+            if move.lot.id not in lot_moves:
+                lot_moves[move.lot.id] = []
+            lot_moves[move.lot.id].append(move)
+
+        for lot in lots:
+            res['total_cost'][lot.id] = Decimal(0)
+            res['cost_price'][lot.id] = Decimal(0)
+            if not lot.id in lot_moves:
+                continue
+
+            for k, v in pbl.items():
+                key = k[1:]
+                if key == (lot.product.id, lot.id):
+                    warehouse_quantity = Decimal(v)
```

Comment:
You're picking the quantity of one of the locations only. If there're more than one warehouse this quantity will be wrong.

You have to sum the quantities of all locations with the given product and lot.